### PR TITLE
Finer-grained locking

### LINKED ISSUE #1266: `No such spec in database` error
This command
```
spack install -j24 cactusext +funhpc +julia +llvm +simulationio %gcc@6.1.0-spack
```
led to this error:
```
==> Installing cactusext
==> openmpi is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/openmpi-2.0.0-5l3pomaqonrx3phvbm4kblzsms56ljtj
==> hwloc is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/hwloc-1.11.3-xgdwohjb2o26uftzaqtcmwf2ejxlhidw
==> git is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/git-2.8.1-jh72meyfuritl2wexqmckrshmufe7az4
==> papi is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/papi-5.4.3-tnovpyennp4cerqns4xnfscfvl6d3zlz
==> hdf5-blosc is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/hdf5-blosc-master-6m7xtyqubxsb2merc36i2oyhtdvno3zp
==> lmod is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/lmod-6.4.1-i2xu3l63aipam7bdlu3ig4eriqtjnbbd
==> hdf5 is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/hdf5-1.10.0-patch1-jfa4zar7ysebuadia7zlhoebrlaeo5lv
==> python is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/python-2.7.11-y6bfneoxof25qt6wgdsbfsra62grfdnz
==> lua is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/lua-5.3.2-xe6niqb4cwgkdpq33qjoi3rby3vcdnew
==> fftw is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/fftw-3.3.4-ojokm7aahpro46e4tvoqgsc57lzbfdum
==> openssl is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/openssl-1.0.2h-ookfh4fsbv5ijqjsewerc5vvf4hzfme6
==> julia is already installed in /home/ux452368/comet/src/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0-spack/julia-master-bzqdzwekzl3geaeyuexssnelbnwfiwsf
Traceback (most recent call last):
  File "/home/ux452368/comet/src/spack/bin/spack", line 179, in <module>
    main()
  File "/home/ux452368/comet/src/spack/bin/spack", line 157, in main
    return_val = command(parser, args)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/cmd/install.py", line 91, in install
    explicit=True)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/package.py", line 943, in do_install
    run_tests=run_tests)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/package.py", line 1096, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/package.py", line 925, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/database.py", line 76, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/database.py", line 461, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/home/ux452368/comet/src/spack/lib/spack/spack/database.py", line 456, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! julia@master%gcc@6.1.0-spack+binutils~cxx+hdf5+mpi~plot~python arch=linux-centos6-x86_64^binutils@2.25%gcc@6.1.0-spack+gold~krellpatch~libiberty arch=linux-centos6-x86_64^bzip2@1.0.6%gcc@6.1.0-spack arch=linux-centos6-x86_64^cmake@3.5.2%gcc@6.1.0-spack~doc+ncurses+openssl~qt arch=linux-centos6-x86_64^curl@7.49.1%gcc@6.1.0-spack arch=linux-centos6-x86_64^expat@2.1.0%gcc@6.1.0-spack arch=linux-centos6-x86_64^git@2.8.1%gcc@6.1.0-spack arch=linux-centos6-x86_64^hdf5@1.10.0-patch1%gcc@6.1.0-spack+cxx~debug+fortran+mpi+shared~szip~threadsafe arch=linux-centos6-x86_64^hwloc@1.11.3%gcc@6.1.0-spack arch=linux-centos6-x86_64^libpciaccess@0.13.4%gcc@6.1.0-spack arch=linux-centos6-x86_64^libsigsegv@2.10%gcc@6.1.0-spack arch=linux-centos6-x86_64^m4@1.4.17%gcc@6.1.0-spack+sigsegv arch=linux-centos6-x86_64^ncurses@6.0%gcc@6.1.0-spack arch=linux-centos6-x86_64^openmpi@2.0.0%gcc@6.1.0-spack~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm+verbs+vt arch=linux-centos6-x86_64^openssl@1.0.2h%gcc@6.1.0-spack arch=linux-centos6-x86_64^python@2.7.11%gcc@6.1.0-spack~ucs4 arch=linux-centos6-x86_64^readline@6.3%gcc@6.1.0-spack arch=linux-centos6-x86_64^sqlite@3.8.5%gcc@6.1.0-spack arch=linux-centos6-x86_64^zlib@1.2.8%gcc@6.1.0-spack arch=linux-centos6-x86_64'
```

This happens on Comet at SDSC; the same Spack command works fine on other systems. How do I debug this?

### PR DESCRIPTION
**Pull request by tgamblin**
Fixes #1266, #1966.

Rebased #866 on `develop`.  See discussion there.

To recap, from @alalazo:

##### Modifications
- [x] removed global db locks from `install`, `uninstall` and `diy`
- [x] added lock for stage directory (an exclusive lock is taken as part of the `Stage` context manager)
- [x] added lock for prefix directory (an exclusive lock is taken during `Package.do_install`)
- [x] debug output shows locks acquire and release
- [x] Use single lock file instead of per-prefix locks.

##### Install and uninstall are not mutually exclusive
This means that if : 
```
spack uninstall ...
``` 
is run concurrently with 
```
spack install ...
```
the former command may cause the latter to fail (the same way an administrator can `sudo rm ...` something that another admin is about to use)

### CODE REVIEW

**Comment by trws** [comment]
This looks good, I will admit though that I'm a bit wary of it without any tests.

**Comment by tgamblin** [comment]
@alalazo @trws: I played around with this a bit and I think a few things are needed.  I'm worried that the lock files:

1. tend to pile up in the install and stage directories (there are a lot of them), and
1. They're all in one directory.

Depending on the parallel FS, having all the locks in one directory can be good or bad, but when it's bad it can be really bad.  So I would like to somehow get the locks _inside_ the stage and prefix directories.  

Given those requirements, I played around with Spack's lock file a bit.  It uses readers-writer locks because it has heretofore only had to deal with locking single files (like the DB) that stick around.  I think that is the right way to lock those files. However, the install and stage directories are transient, and they'll always need an exclusive lock (not reader locks).  

I think the best thing to do would be to use simple exclusive file locks (`open` with `O_EXCL`) to control these directories.  They have the advantage that the files only exist when in use, and I think we could implement them such that they live inside the install prefix and/or stage.  I can't do that with reader-writer locks.  RW locks don't play nicely with being deleted -- you get races between open and the first call to `fcntl` that I don't think are possible to get rid of when the lock files are deleted on release.

So, if you guys aren't opposed, I would like to take a stab at introducing a new exclusive file lock (pretty simple) and using that _inside_ the stage and prefix directories where @alalazo currently uses the current exclusive `fcntl` locks.

Thoughts?

**Comment by alalazo** [comment]
> So, if you guys aren't opposed, I would like to take a stab at introducing a new exclusive file lock (pretty simple) and using that inside the stage and prefix directories where @alalazo currently uses the current exclusive fcntl locks.

:+1:  This will surely be an improvement over the current implementation

**Comment by trws** [comment]
That sounds like a good idea to me, as long as we can count on an outer lock to prevent creation of multiple versions of an otherwise redundant stage directory.

**Comment by tgamblin** [comment]
@alalazo: this is reworked.  Can you test it out?

The approach is slightly different (but better IMHO) than what I suggested above.  I figured out how we can have a single lock file for all the prefixes and a single file for all the stages and ALSO get readers-writer semantics.

How it works for prefixes:

1. The lock file is in `opt/spack/.spack-db/prefix_lock` and it lives alongside the install DB.

1. Each lock is an `fcntl` byte-range lock on the _n_th byte of a file.  Byte range locks can extend past the end of a file w/o taking up space, so we can have up to `2^63-1` of these on 64-bit machines.

1. _n_ is the `sys.maxsize`-bit prefix of the DAG hash.

Stages are similar -- they use a prefix of the `sha1` of the stage name instead of the DAG hash prefix.

With 100 concurrent builds, the likelihood of a lock collision (false positive) is ~5.36e-16, so this scheme should retain more than sufficient parallelism (with no chance of false negatives), and get us reader-writer lock semantics with a single file, so no need to clean up lots of lock files.

**Comment by tgamblin** [comment]
Also this fixes #1266.

**Comment by tgamblin** [comment]
Also, if anyone is interested, both of these are good reads:
* [Everything you never wanted to know about file locking](http://apenwarr.ca/log/?m=201012#13)
* [On the Brokenness of File Locking(http://0pointer.de/blog/projects/locking.html)

TL; DR: `fcntl` locks (`lockf` in Python... sigh)  work on NFS v2.something and up, and most sites deploy NFS v3 or v4, where they work now.  They also work on GPFS and Lustre.  I _believe_ that covers most filesystems we care about.  These locks _should_ let us do parallel builds, and the readers-writer semantics should be useful for things like taking read locks out on dependencies while packages are being installed (preventing others from uninstalling them while that happens and before the new package that depends on them is in the DB).

**Comment by alalazo** [comment]
@tgamblin Thanks for the references above!

**Comment by alalazo** [comment]
> Each lock is an fcntl byte-range lock on the nth byte of a file. Byte range locks can extend past the end of a file w/o taking up space, so we can have up to 2^63-1 of these on 64-bit machines.

Quite clever! I didn't know you could do something like this.

**Comment by tgamblin** [comment]
@alalazo: I still wish there was zero chance of taking a lock unnecessarily but sha1 collision likelihood is very low even for a prefix.

If this works for you I'll merge.  Note that this also fixes a bug with the backported read only lock fix.

**Comment by tgamblin** [comment]
@adamjstewart @davydden: I need someone besides me to test this -- if either of you have time can you verify it fixes #1266?

**Comment by adamjstewart** [comment]
Will do.

**Comment by adamjstewart** [comment]
I'm seeing some problems on my end:
```
[01:59:45] ajstewart@blogin4:/soft/spack-0.9.1 (develop)
$ git co features/db-locking
Branch features/db-locking set up to track remote branch features/db-locking from upstream.
Switched to a new branch 'features/db-locking'
[01:59:52] ajstewart@blogin4:/soft/spack-0.9.1 (features/db-locking)
$ spack edit m4
[02:00:14] ajstewart@blogin4:/soft/spack-0.9.1 (features/db-locking *)
$ spack install gmp
==> gmp is already installed in /blues/gpfs/home/software/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/gmp-6.1.1-e3x7lcgg4qizghb53izi72aep6ybswvp
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 192, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/install.py", line 97, in install
    explicit=True)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 981, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 79, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 527, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
[02:00:24] ajstewart@blogin4:/soft/spack-0.9.1 (features/db-locking *)
$ spack reindex
[02:01:28] ajstewart@blogin4:/soft/spack-0.9.1 (features/db-locking *)
$ spack install gmp
==> gmp is already installed in /blues/gpfs/home/software/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/gmp-6.1.1-e3x7lcgg4qizghb53izi72aep6ybswvp
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 192, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/install.py", line 97, in install
    explicit=True)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 981, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 79, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 527, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```

**Comment by alalazo** [comment]
@tgamblin I'll have a chance to test this tomorrow...

**Comment by tgamblin** [comment]
@adamjstewart: can you attach output of `spack debug create-db-tarball`?

**Comment by adamjstewart** [comment]
Um, I'm not sure if I'm doing this right...
```
$ spack debug create-db-tarball > create-db-tarball.txt
tar (child): /blues/gpfs/home/software/spack-0.9.1/spack-db.features/db-locking.ca5f5ee.2016-10-06-160940.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
/bin/tar: /blues/gpfs/home/software/spack-0.9.1/spack-db.features/db-locking.ca5f5ee.2016-10-06-160940.tar.gz: Wrote only 4096 of 10240 bytes
/bin/tar: Error is not recoverable: exiting now
==> Error: Command exited with status 2:
```

**Comment by tgamblin** [comment]
@adamjstewart: `spack debug create-db-tarball` just creates a tarball.  I meant to attach the tarball.  Sorry that was ambiguous.

**Comment by adamjstewart** [comment]
@tgamblin If it creates a tarball, I didn't see one. It just spit out thousands of lines of text for me. So I tried writing them to a file, but that crashed. Let me try again on develop and then on your branch.

**Comment by adamjstewart** [comment]
Now I'm getting completely different behavior. On develop:
```
$ spack debug create-db-tarball
==> Created spack-db.develop.c2ca869.2016-10-10-152307.tar.gz
```
I deleted that file and then immediately checked out your branch:
```
$ spack debug create-db-tarball
tar: Failed to open '/Users/Adam/spack/spack-db.features/db-locking.ca5f5ee.2016-10-10-152419.tar.gz'
==> Error: Command exited with status 1:
'/usr/bin/tar' '-czf' '/Users/Adam/spack/spack-db.features/db-locking.ca5f5ee.2016-10-10-152419.tar.gz' 'opt/spack/.spack-db/index.yaml'
```

**Comment by adamjstewart** [comment]
This is with nothing installed btw.

**Comment by adamjstewart** [comment]
But, I tried following [my steps to reproduce the bug](https://github.com/LLNL/spack/issues/1266#issuecomment-240807169) and it didn't occur this time, so that's promising!

**Comment by tgamblin** [comment]
@adamjstewart: D'oh.  Forgot to filter the symbolic name in that command.  Rebased and filtered out illegal filename characters (like `/`). 

Can you try now?

**Comment by tgamblin** [comment]
> But, I tried following my steps to reproduce the bug and it didn't occur this time, so that's promising!

That is promising.  I think your error might be coming from something else -- I just want to confirm with the DB test data.

**Comment by adamjstewart** [comment]
@tgamblin Good catch! `spack debug create-db-tarball` works for me again. 

Although [my steps to reproduce the bug](https://github.com/LLNL/spack/issues/1266#issuecomment-240807169) can no longer reproduce the bug on my laptop (macOS), I'm still seeing a problem on our CentOS 6 cluster. Here is the output:
```
$ spack install gmp %pgi
==> Installing gmp
==> Installing m4
==> libsigsegv is already installed in /blues/gpfs/home/software/spack-0.9.1/opt/spack/linux-centos6-x86_64/pgi-16.5-0/libsigsegv-2.10-3zrk2bhqrgknreumylyvff2tzpqlrt2d
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 192, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/install.py", line 97, in install
    explicit=True)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1007, in do_install
    dirty=dirty)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1007, in do_install
    dirty=dirty)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 985, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 79, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 527, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```
And here is my debug tarball:
[spack-db.features-db-locking.842c36c.2016-10-10-175746.tar.gz](https://github.com/LLNL/spack/files/520577/spack-db.features-db-locking.842c36c.2016-10-10-175746.tar.gz)

**Comment by adamjstewart** [comment]
Aside from the obvious OS difference, the biggest difference is that on my laptop, I didn't have anything installed before trying out your branch, whereas on our cluster, I had 460 packages already installed (mostly duplicates). That assert statement ought to shed some light on the problem though.

**Comment by tgamblin** [comment]
@adamjstewart: I looked into this.  The cause for your issue is separate from what this issue is trying to fix so I made a separate issue for it (#1992) and threw it in the 0.10 release.
